File: scripts/run_capture_batch.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
import json
import subprocess
import sys
from pathlib import Path


ROOT = Path(__file__).resolve().parent.parent


def run_preflight() -> None:
    result = subprocess.run(
        [sys.executable, str(ROOT / "scripts" / "check_capture_env.py")],
        text=True,
        capture_output=True,
        cwd=ROOT,
    )
    if result.returncode != 0:
        if result.stdout:
            print(result.stdout, end="")
        if result.stderr:
            print(result.stderr, end="", file=sys.stderr)
        raise RuntimeError("Capture environment preflight failed.")
    if result.stdout:
        print(result.stdout, end="")


def load_config(path: Path) -> dict:
    return json.loads(path.read_text(encoding="utf-8"))


def capture_site(
    batch: str,
    slug: str,
    url: str,
    desktop_width: int,
    desktop_height: int,
    mobile_device: str,
) -> int:
    cmd = [
        sys.executable,
        str(ROOT / "scripts" / "capture_site.py"),
        "--batch",
        batch,
        "--slug",
        slug,
        "--url",
        url,
        "--desktop-width",
        str(desktop_width),
        "--desktop-height",
        str(desktop_height),
        "--mobile-device",
        mobile_device,
    ]
    result = subprocess.run(cmd, text=True, capture_output=True, cwd=ROOT)
    if result.stdout:
        print(result.stdout, end="")
    if result.stderr:
        print(result.stderr, end="", file=sys.stderr)
    return result.returncode


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Run a capture batch using agent-browser."
    )
    parser.add_argument("--config", required=True, help="Path to batch config JSON")
    parser.add_argument("--desktop-width", type=int, default=1440)
    parser.add_argument("--desktop-height", type=int, default=900)
    parser.add_argument("--mobile-device", default="iPhone 14")
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    try:
        config_path = (
            (ROOT / args.config).resolve()
            if not Path(args.config).is_absolute()
            else Path(args.config)
        )
        config = load_config(config_path)
        batch = config["batch"]
        sites = config["sites"]
        run_preflight()
    except Exception as error:  # noqa: BLE001
        print(str(error), file=sys.stderr)
        return 1

    failures: list[str] = []
    for site in sites:
        slug = site["slug"]
        url = site["url"]
        print(f"\n==> Capturing {slug} ({url})")
        exit_code = capture_site(
            batch=batch,
            slug=slug,
            url=url,
            desktop_width=args.desktop_width,
            desktop_height=args.desktop_height,
            mobile_device=args.mobile_device,
        )
        if exit_code != 0:
            failures.append(slug)

    if failures:
        print(
            "\nCapture batch completed with failures:",
            ", ".join(failures),
            file=sys.stderr,
        )
        return 1

    print(f"\nCapture batch {batch} completed successfully.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/run-capture-batch.ts
#!/usr/bin/env bun

import { spawnSync } from "node:child_process";
import { readFileSync } from "node:fs";
import { isAbsolute, resolve } from "node:path";

const ROOT = resolve(import.meta.dir, "..");

type BatchConfig = {
  batch: string;
  sites: Array<{ slug: string; url: string }>;
};

function parseArgs(argv: string[]): Record<string, string> {
  const args: Record<string, string> = {};

  for (let i = 0; i < argv.length; i += 1) {
    const key = argv[i];
    if (!key.startsWith("--")) continue;
    const value = argv[i + 1];
    if (!value || value.startsWith("--")) {
      throw new Error(`Missing value for ${key}`);
    }
    args[key.slice(2)] = value;
    i += 1;
  }

  return args;
}

function run(cmd: string[]): ReturnType<typeof spawnSync> {
  return spawnSync(cmd[0], cmd.slice(1), {
    cwd: ROOT,
    encoding: "utf8"
  });
}

const args = parseArgs(process.argv.slice(2));
const configArg = args.config;
const desktopWidth = args["desktop-width"] || "1440";
const desktopHeight = args["desktop-height"] || "900";
const mobileDevice = args["mobile-device"] || "iPhone 14";

if (!configArg) {
  console.error("Usage: bun run scripts/run-capture-batch.ts --config <config-path>");
  process.exit(1);
}

const configPath = isAbsolute(configArg) ? configArg : resolve(ROOT, configArg);
const config = JSON.parse(readFileSync(configPath, "utf8")) as BatchConfig;

const preflight = run(["bun", "run", "scripts/check-capture-env.ts"]);
if (preflight.stdout) process.stdout.write(preflight.stdout);
if (preflight.stderr) process.stderr.write(preflight.stderr);
if (preflight.status !== 0) {
  console.error("Capture environment preflight failed.");
  process.exit(1);
}

const failures: string[] = [];

for (const site of config.sites) {
  console.log(`\n==> Capturing ${site.slug} (${site.url})`);
  const result = run([
    "bun",
    "run",
    "scripts/capture-site.ts",
    "--batch",
    config.batch,
    "--slug",
    site.slug,
    "--url",
    site.url,
    "--desktop-width",
    desktopWidth,
    "--desktop-height",
    desktopHeight,
    "--mobile-device",
    mobileDevice
  ]);

  if (result.stdout) process.stdout.write(result.stdout);
  if (result.stderr) process.stderr.write(result.stderr);
  if (result.status !== 0) failures.push(site.slug);
}

if (failures.length > 0) {
  console.error(`\nCapture batch completed with failures: ${failures.join(", ")}`);
  process.exit(1);
}

console.log(`\nCapture batch ${config.batch} completed successfully.`);
